### ISSUE TITLE
Dry-run exits with trailing_stop loss tag at price higher than custom_stoploss()

### ISSUE THREAD

**Issue by GeorgeGkinis** [freqtrade/freqtrade#12462]
## Describe your environment

Operating System:       Linux-6.14.0-34-generic-x86_64-with-glibc2.39
Python Version:         Python 3.12.3
CCXT Version:           4.5.12

Freqtrade Version:      freqtrade 2025.10
  
## Your question

I have defined custom_stoploss() to return a level according to an indicator (the yellow line in the screenshot that is actually an ema):

```python
    def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime,
                        current_rate: float, current_profit: float, after_fill: bool,
                        **kwargs) -> float | None:
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        candle = dataframe.iloc[-1].squeeze()
        return stoploss_from_absolute(candle["stop_ref"],
                                      current_rate=current_rate,
                                      is_short=trade.is_short,
                                      leverage=trade.leverage)
```

When in dry-run mode I often see trades exiting at a price higher than the indicator.

<img width="1240" height="879" alt="Image" src="https://github.com/user-attachments/assets/9eebf032-30db-47b9-99a9-f65766afa94a" />

I wonder if this has to do with any of the following:
- I increase my position using adjust_trade_position and the stoploss calculates things from the adjusted price
- I should use **current_rate=candle['low']** instead of **current_rate=current_rate**
- I incorrectly set candle as the current open candle instead of the last closed one
- The price has rebounded back up from the moment the stoploss was hit untill the exit order was filled, although I should see the low price wick under the stoploss.

The red line of the stoploss does follow the indicator level (screenshot from another open trade) but exits at prices above that.

<img width="680" height="806" alt="Image" src="https://github.com/user-attachments/assets/433a4b57-3409-4790-8734-7b2efdbde88e" />

Here is another example of a trade that turned red when it would have been in profit had it not exited due to the wrong stop_loss:

<img width="1314" height="1318" alt="Image" src="https://github.com/user-attachments/assets/2335765b-9b4e-4993-bf50-28e688d4c221" />

My properties file is like this:
```json
{
  "strategy_name": "PumpyPyramidCSL",
  "params": {
    "roi": {},
    "stoploss": {
      "stoploss": -0.9
    },
    "trailing": {
      "trailing_stop": false,
      "trailing_stop_positive": 0.148,
      "trailing_stop_positive_offset": 0.197,
      "trailing_only_offset_is_reached": false
    },
    "max_open_trades": {
      "max_open_trades": 1000
    },
    "buy": {
      "quant_lvl_enter": 0.99,
      "quant_lvl_exit": 0.03,
      "quant_win": 200,
      "stake_pct": 0.3,
      "adx_lvl": 20,
      "adx_win": 25,
      "ema_fast": 30,
      "ema_slow": 450,
      "kama_enter": 30,
      "kama_exit": 25,
      "lin_fast": 25,
      "lin_kama": 30,
      "lin_slow": 30,
      "min_volume": 1000
    },
    "sell": {},
    "protection": {}
  },
  "ft_stratparam_v": 1,
  "export_time": "2025-11-02 00:00:19.591567+00:00"
}
```

The config file used:

```json
{
  "exchange": {
    "name": "kraken",
    "use_public_trades": false,
    "key": "no-need-to-know",
    "secret": "this-is-secret",
    "timeout": 60000,
    "ccxt_config": {
      "enableRateLimit": true,
      "rateLimit": 100
    },
    "ccxt_async_config": {
      "enableRateLimit": true,
      "rateLimit": 100
    },
    "asyncio_loop": true,
    "ccxt_rate_limit": true,
    "pair_whitelist": [
      ".*/USD"
    ],
    "pair_blacklist": [
      "USD.*",
      "EUR.*"
    ]
  },
  "stoploss_on_exchange": true,
  "trading_mode": "spot",
  "order_time_in_force": {
    "entry": "GTC",
    "exit": "GTC"
  },
  "entry_pricing": {
    "price_side": "other",
    "use_order_book": true,
    "order_book_top": 5,
    "check_depth_of_market": {
      "enabled": true,
      "top": 1
    }
  },
  "exit_pricing": {
    "price_side": "other",
    "use_order_book": true,
    "order_book_top": 5,
    "check_depth_of_market": {
      "enabled": true,
      "top": 1
    }
  },
  "pairlists": [
    {
      "method": "StaticPairList"
    },
    {
      "method": "VolumePairList",
      "number_assets": 100,
      "sort_key": "quoteVolume",
      "min_value": 500,
      "refresh_period": 300,
      "lookback_timeframe": "5m",
      "lookback_period": 4
    }
  ],
  "stake_currency": "USD",
  "max_open_trades": 1000,
  "dry_run_wallet": 1000,
  "stake_amount": "unlimited",
  "dry_run": true,
  "tradable_balance_ratio": 0.99,
  "fiat_display_currency": "USD",
  "margin_mode": "isolated",
  "timeframe": "5m",
  "cancel_open_orders_on_exit": false,
  "api_server": {
    "enabled": true,
    "listen_ip_address": "0.0.0.0",
    "listen_port": 8080,
    "verbosity": "info",
    "enable_openapi": false,
    "jwt_secret_key": "whatever",
    "ws_token": "whatever",
    "CORS_origins": [
      "http://0.0.0.0:8083",
      "http://q-nuc:8083",
      "http://0.0.0.0:8080",
      "http://q-nuc:8080"
    ],
    "username": "user",
    "password": "pass"
  },
  "bot_name": "PumpyPyramidKraken",
  "initial_state": "running",
  "force_entry_enable": false,
  "internals": {
    "process_throttle_secs": 5
  }
}
```

Any suggestion about what i am doing wrong please?
Thanks :)

**Comment by xmatthias** [comment]
`"stoploss_on_exchange": true,`does not exist at the top level - so that config setting is irrelevant

Also the parameters file you shared is pretty much useless - as that's only relevant with the strategy at hand (and even then you're not using parameters for your stop function - so . . .

Your "stoploss line" confirms that your calculation seems to match - and i _assume_  that's also the case when you're adjusting positions.


What's missing (from an information perspective) is the order types configuration - which i think is pretty relevant here.

> The price has rebounded back up from the moment the stoploss was hit untill the exit order was filled, although I should see the low price wick under the stoploss.

I think that's pretty much exactly (one of the) problems. 
stoploss order triggering/matching happens on orderbook pricing.
the bid of the orderbook however doesn't have to correspond to the low of the candle - so you might not see this in a dry-run - as price recovered (bids WERE below your stop - but your fill isn't).

In a non-dry setup - you'd be the one causing the (new low) to materialize (independent of size, really) - but in dry-run you have no impact on the market - so there's nothing showing on the chart in this regard.

**Comment by GeorgeGkinis** [comment]
Thanks for your answer @xmatthias!

> "stoploss_on_exchange": true,does not exist at the top level - so that config setting is irrelevant

thanks, that slipped me!

> Also the parameters file you shared is pretty much useless ...

I provided it to indicate that `trailing_stop` is set to `false` and and `stoploss` to -0.9 and as such not the reasonsfor exiting above the 'custom_stoploss' price.

> What's missing (from an information perspective) is the order types configuration - which i think is pretty relevant here.

I have added the following that hopelly fixes things:
```json
  "order_types": {
    "entry": "limit",
    "exit": "limit",
    "emergency_exit": "market",
    "force_exit": "market",
    "force_entry": "market",
    "stoploss": "limit",
    "stoploss_on_exchange": true,
    "stoploss_on_exchange_interval": 60
  },
```
I have to wait until the market changes so that new trades are placed.

I will make a variant of the strategy that uses `custom_exit() ` with the same level and let it run alongside the current one. Then review the exits of the same trades to see if I will have similar behaviour and as such if the reason for the exit price discrepancy has to do with my settings, a creepy bug or the orderbook depth as you mentioned.

I will not outrule a potential bug at this moment as there is no visual way to check at which level the custom_stoploss() was historically.
That feature was discussed in #12230

I will let you know in a couple of days.

**Comment by xmatthias** [comment]
well if any, it'd be a bug in your strategy logic (like, the sl column is repainting or similar).

As said above - stop triggers happen on the orderbook - based on your configured pricing logic - with the fill also happening on the orderbook.

Also - by default, you supply one price (the trigger price) - which is then multiplied with 0.99 (the default for stoploss_on_exchange_limit_ratio) - which constitutes the limit order.
essentially this means that the stop sell order will place result in a limit order with this price as limit price.
That's not exactly your problem - but still kinda related - as a stoploss order fills identically to a regular order, which will start at the best market bid (but probably checked a little later - as dry-run does a couple more things between stop and fill)

To have something to look at properly - you'll anyway want logs, not screenshots, really.
these _could_  contain some more info - but maybe don't - as you'll still have the (volatile) prices of the orderbook, which you can't really retrace after the fact.

**Comment by GeorgeGkinis** [comment]
Using custom_exit(): (use_custom_stoploss = False )

```python
    def custom_exit(
            self,
            pair: str,
            trade: Trade,
            current_time: datetime,
            current_rate: float,
            current_profit: float,
            **kwargs
    ):
        # Get latest analyzed dataframe
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        if dataframe is None or len(dataframe) == 0:
            return None
        last_candle = dataframe.iloc[-1].squeeze()

        if current_rate < last_candle['stop_ref']:
            return f'exit_stoploss'
        return None
```

Actual exit price is lower than the indicator as expected when using the orderbook.
I believe it is called slippage.

<img width="928" height="913" alt="Image" src="https://github.com/user-attachments/assets/5c8c2329-08a4-41f0-87e8-e50459383741" />

Using custom_stoploss(): (    use_custom_stoploss = True)
```python
    def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime,
                        current_rate: float, current_profit: float, after_fill: bool,
                        **kwargs) -> float | None:
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()
        return stoploss_from_absolute(last_candle["stop_ref"],
                                      current_rate=current_rate,
                                      is_short=trade.is_short,
                                      leverage=trade.leverage)
```
Exit price many % above the expected level.
I have yet to see any exit from custom_stoploss below the indicator.
<img width="681" height="862" alt="Image" src="https://github.com/user-attachments/assets/32f8177e-2ea5-4662-abb9-8103e0c8449f" />

In both cases the comparison is between the current_rate and the stop_ref indicator

custom_stoploss() exits at an higher price and it is driving me nuts :)

custom_exit works like a charm!

The parameters file was not the same :/
I will compare apples-to-apples next.

**Comment by xmatthias** [comment]
you never were comparing apples to apples - what you need to look at is logs (probably with additional output added from the callbacks - like the value of the stoploss for that candle) - not screenshots.

From your screenshots, it's pretty likely that your indicator has repainting / lookahead bias issues (both up and down) - so the value is different while the candle is the most recent - but when you look at it a couple minutes/hours later, the value changed - making it seem like the value was different and the exit was "odd".

**Comment by GeorgeGkinis** [comment]
I am not sure how indicator repainting can be achieved in my case.
This is my indicator snippet:
```
       def populate_indicators(self, df: pd.DataFrame, metadata: dict) -> pd.DataFrame:

            df['ema_fast'] = ta.EMA(df['low'], timeperiod=self.ema_fast.value)

            # Dynamic stop anchor
            df['stop_ref'] = df['ema_fast']
```
btw I now have the same parameter file for both strategies running side-by-sideand I still see the same behavior in multiple occasions. 
custom_stoploss() exits way above the indicator, custom_exit() just below it due to slippage and that is acceptable.

If there was an indicator redrawing it should happen in both cases.

I do agree that some logging can be added.
I can log the indicator value, current_rate and the orderbook at the exit point in custom_stoploss() and custom_exit() when they are triggered.

But no more screenshots I promise ;)

**Comment by xmatthias** [comment]
Will be closing this assuming that you've found the problem in your code considering that we didn't hear from you in 3 weeks.

Feel free to comment below if there's still something unclear on this topic so we can reopen or at least answer :)
